✅ Playwright e2e tests

KemingHe opened the pull request:
# Introducing Accessibility and Playwright to Socialify

_Testing and aria-label only changes._ I.e. in `repo.tsx`:

```Typescript
<input
  id="repo-input"
  name="repo-input"
  className="input flex-1 pl-3 font-bold bg-base-200 focus:outline-none"
  ref={repoInputRef}
  type="text"
  value={repoInput}
  onChange={(e) => {
    setRepoInput(e.target.value)
  }}
  aria-required="true"
  aria-label="GitHub repository"
  required
/>
```

@wei Please review this after #403 , really appreciated!

## What Changed

* Scripts in `package.json` now reflects unit and e2e testing:

```json
"test:unit": "jest",
"test:unit:watch": "jest --watch",
"test:unit:update-snapshot": "jest -u",
"test:e2e": "playwright test",
"test": "yarn test:unit && yarn test:e2e",
```

* E2e tests defined in `.playwright/` and ignored by Jest.
* E2e **screenshots (NOT snapshots)** generated per `**.spect.ts`.

## Why

I'm working on the DaisyUI v2->v4 styling upgrade and I need an automated e2e testing framework to make sure that **even though the `tsx` is changing due to the upgrade, the user stories/app behavior remain the same**. This is critical to ensuring the upgrade is styling-only.

## Further Why

We currently only have Jest snapshot and unit tests, which do not cover the behavior of the app. Plus, the Jest snapshots will be updated once the styling change starts to happen.

wei reviewed (changes requested):
@KemingHe Excellent work, I don't seem to have write permission to your fork/branch. could you help with a few minor changes to get the tests to pass?

- [ ] Let's just run playwright on one browser - Chrome for now to save disk space and time.
- [ ] Please add the some basic documentation on this in README or CONTRIBUTING.md or where appropriate
- [ ] Please update `.github/workflows/build.yml` to set the environment variables and install the necessary playwright browser so GitHub Actions run the test successfully, something like:

    ```
    env:
      GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
    ```

wei commented on playwright.config.ts:
```diff
@@ -0,0 +1,109 @@
+
+  // Run your local dev server before starting the tests.
+  webServer: {
+    command: 'yarn dev',
```
Shall we use yarn start for testing instead of starting the development server?

KemingHe replied:
Okay, let's do that, I would further use `yarn build && yarn start` as the command to make sure we are starting the latest prod build.

wei commented on src/components/repo/repo.tsx:
```diff
@@ -38,22 +38,34 @@ const Repo: React.FC = () => {
                 <div className="form-control">
                   <div className="input-group">
-                    <span className="pr-0 bg-base-200">
+                    <label htmlFor="repo-input" className="sr-only">
```
What does sr-only mean?

wei replied:
Oh screen reader only? Could you help me understand the benefit of it and why this is the only instance of this class?

KemingHe replied:
Hi, Wei, yes: `className="sr-only"` means the element remains invisible to regular users while accessible for screen reader softwares. 

You would often see this used on invisible input checkboxes that toggles hamburger menus.

I used it here because the repo input box lacked a label. This is not an issue for most able-bodied people, but again, we are trying to make the site accessible to all.

KemingHe commented:
Hi @wei, I have some exciting updates and new proposals for you:

>  [ ] Let's just run playwright on one browser - Chrome for now to save disk space and time.

**I propose we keep both `Chromium` and `Mobile Chrome` as this covers responsiveness of the app while keeping our footprint small.**

> [x]  Perhaps remove the stargazers for now so the tests don't fail due to changes in star count.

Done in 3488c71 where I checked out only the updated Playwright test and config from the 100% working #405 branch.

I will keep you posted on the rest of the requests. ☀️

KemingHe commented:
> - [x]  Please add the some basic documentation on this in README or CONTRIBUTING.md or where appropriate

@wei This is fully delivered in 4a4724b, also added the proposed list of contributors from #407, as a lean alternative to the comment in #405 about ballooning up `package.json`.

> [!NOTE]
> I added all the instructions to README instead of CONTRIBUTING.md because technical setup guides changes often, requires high visibility; plus CONTRIBUTING.md (content-wise) is more about Code of Conduct, rather than how-tos.

KemingHe commented:
Now onto the GitHub Action scripts...

KemingHe commented:
>  I don't seem to have write permission to your fork/branch. could you help with a few minor changes to get the tests to pass?

Also just invited you as a collaborator to `KemingHe/contrib-socialify`. Maybe a bit unnecessary as our current routine works fine. Just FYI.

wei commented:
> >  I don't seem to have write permission to your fork/branch. could you help with a few minor changes to get the tests to pass?
> 
> 
> 
> Also just invited you as a collaborator to `KemingHe/contrib-socialify`. Maybe a bit unnecessary as our current routine works fine. Just FYI.

Sounds good! I meant this: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

KemingHe commented:
> > > I don't seem to have write permission to your fork/branch. could you help with a few minor changes to get the tests to pass?
> > 
> > 
> > Also just invited you as a collaborator to `KemingHe/contrib-socialify`. Maybe a bit unnecessary as our current routine works fine. Just FYI.
> 
> Sounds good! I meant "Allow edits from maintainers" checkbox: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

Oh, thanks for teaching me about this. I thought it has been checked this whole time? And is there a way for you to toggle it on by default for all PRs to Socialify?

<img width="322" alt="Screenshot 2024-12-11 at 10 18 24 AM" src="https://github.com/user-attachments/assets/e0098e20-a8bd-48a5-a22a-0147ff5cc5a8">

<hr/>

And I assume our agreement on:

- `yarn dev` -> `yarn build && start` Playwright config discussion
- using `className="sr-only" discussion

such that I can mark them as resolved?

wei commented:
> And I assume our agreement on:
> 
> yarn dev -> yarn build && start Playwright config discussion
> using `className="sr-only" discussion
> such that I can mark them as resolved?

LGTM!

wei commented:
Excellent @KemingHe ! Quick thing - If you don't mind, let's just keep one changeset file for each PR. 
You can add multiple lines in each changeset file instead of separate files. You can edit them manually, just need 1 file with all the changes per PR so we don't bump multiple minor and patch versions from merging a single PR~

KemingHe commented:
> Excellent @KemingHe ! Quick thing - If you don't mind, let's just keep one changeset file for each PR. You can add multiple lines in each changeset file instead of separate files. You can edit them manually, just need 1 file with all the changes per PR so we don't bump multiple minor and patch versions from merging a single PR~

Understood, so for this PR as an example would it look like:

```markdown
---
"socialify": minor
---

Added Playwright e2e testing to workflow.

Updated Playwright tests and added setup instructions in README.md

Description of other updates.

... 
```

**Two Questions:**

1. Are we marking the `"socialify": version` by the greatest since we can only have one version (`major > minor > patch`)?
2. Are we adding descriptions in chrono order (oldest first/top of file) or reverse?

KemingHe commented:
> @KemingHe exactly!
> 
> 1. Yes, minor is good
> 2. Doesn't matter too much. The commit history and PR will have more detailed information of exactly what's included so you can just include more of a summary / overview. A few words or a general description of the PR / release will do.

I will also add this detail to the `README.md` once I finish debugging `build.yml`. No worries.

KemingHe commented:
> I will also add this detail to the README.md once I finish debugging build.yml. No worries.

![Screenshot from 2024-12-11 12-45-36](https://github.com/user-attachments/assets/7015676a-f93d-4c12-a190-5618142822d6)

KemingHe commented:
> @KemingHe Excellent work, I don't seem to have write permission to your fork/branch. could you help with a few minor changes to get the tests to pass?
> 
> ...
> * [ ]  Please update `.github/workflows/build.yml` to set the environment variables and install the necessary playwright browser so GitHub Actions run the test successfully, something like:
>   ```
>   env:
>     GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
>   ```
> * [ ]  Run the build.yml workflow in your fork to ensure it completes successfully~


Hello @wei, here's the latest roadblock that requires significant changes to the project:

* We haven't yet explicitly defined a custom font for Socialify.
* Thus, each browser uses their own default font to render the text.
* Thus, the Playwright screenshots will always fail.
* Thus, we need to add specific font imports in `pages/_app.tsx`.

What do you think?

KemingHe commented:
### As for 

> * [ ]  Please update `.github/workflows/build.yml` to set the environment variables and install the necessary playwright browser so GitHub Actions run the test successfully, something like:
>   ```
>   env:
>     GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
>   ```

Here's a quick update to save you time looking through the code, via snippet from `build.yml`:

```yaml
- name: Lint, build, and test Socialify
  run: yarn verify
  env:
  # GITHUB_TOKEN is a reserved env, use step-level rewrite to
  # temporarily replace with Socialify's repo-scoped GITHUB_TOKEN env.
  GITHUB_TOKEN: ${{ secrets.SOCIALIFY_GITHUB_TOKEN }}

  # For uniformity, also add the project URL as a secret,
  # with the same SOCIALIFY_ prefix.
  PROJECT_URL: ${{ secrets.SOCIALIFY_PROJECT_URL }}
```

Remember to add the two repo-level secrets to `wei/socialify`. 🌻

wei commented:
Good catch!! We can add a default font for sure. Let's pick one that looks similar to the default and explicitly define it (import from Google fonts etc if needed)

Can we just use the default secrets.GITHUB_TOKEN and a default   PROJECT_URL so they don't need to be specified in repo settings?

KemingHe commented:
I can quickly implement the font imports and keep the repo secret config if that's okay with you.

KemingHe commented:
As a stretch goal, I can also make the Playwright artifact export conditional only on failure and add verbosity so devs don't have to fish it out of the console output. They would just be presented the link when viewing the run page.

wei commented:
Ok, let’s try Roboto!

Could you also make sure the pipeline fails when the build fails, for some reason it’s still green when yarn verify failed with a non-zero code: https://github.com/wei/socialify/actions/runs/12282889906/job/34275125170

Let’s just hard code the `PROJECT_URL` to `http://localhost:3000`, and can we use the built-in `GITHUB_TOKEN` to access `wei/socialify` itself in the test? if not, let’s use {{ secrets.PAT }} for now since that is already a secret used in other workflows in the repo.

KemingHe commented:
> Ok, let’s try Roboto!

Great, on it!

> Could you also make sure the pipeline fails when the build fails, for some reason it’s still green when yarn verify failed with a non-zero code: https://github.com/wei/socialify/actions/runs/12282889906/job/34275125170

Yes, of course, the current hacky setup is to always export Playwright report. Def will change that before final review.

> Let’s just hard code the `PROJECT_URL` to `http://localhost:3000`, and can we use the built-in `GITHUB_TOKEN` to access `wei/socialify` itself in the test? if not, let’s use {{ secrets.PAT }} for now since that is already a secret used in other workflows in the repo.

Understood, will change to:

```
- name: Lint, build, and test Socialify
  run: yarn verify
  env:
  # GITHUB_TOKEN is a reserved env, use step-level rewrite to
  # temporarily replace with Socialify's repo-scoped GITHUB_TOKEN env.
  GITHUB_TOKEN: ${{ secrets.PAT }}
  PROJECT_URL: "http://localhost:3000"
```

wei commented on .github/workflows/build.yml:
```diff
@@ -32,5 +32,32 @@ jobs:
+      - name: Lint, build, and test Socialify
+        run: yarn verify
+        # Continue on error to ensure the Playwright artifacts are uploaded.
+        continue-on-error: true
```
This is causing the pipeline to return success even though it failed. Perhaps we need to remove or add another step to fail the job?

wei replied:
Saw your comment above. TY!

KemingHe replied:
Yes, this is a temp hack. Will def change it. It's for testing on my experimental branch `test-addedplaywrightci-keminghe`.

KemingHe commented:
Hi, @wei , sorry just realized that the `<label className="sr-only"...` was breaking the UI such that it got ride of the rounded corners. **I'm thus removing it and falling back to simply adding an accessible `aria-lable="GitHub repository"` to the corresponding `input`:

```Typescript
<input
  id="repo-input"
  name="repo-input"
  className="input flex-1 pl-3 font-bold bg-base-200 focus:outline-none"
  ref={repoInputRef}
  type="text"
  value={repoInput}
  onChange={(e) => {
    setRepoInput(e.target.value)
  }}
  aria-required="true"
  aria-label="GitHub repository"
  required
/>
```

KemingHe commented:
Fixed CI false negative and added nice output window to download artifact instead of manually fishing for it in the console, for example: 

https://github.com/KemingHe/contrib-socialify/actions/runs/12286075397

![Screenshot from 2024-12-11 17-43-25](https://github.com/user-attachments/assets/0a442488-53a6-4b4b-bbee-214d40876ba1)

KemingHe commented:
Font is now confirmed to be the cause for Playwright failure:

### Before

```shell


Edit: typo

KemingHe commented:
Hi @wei , we have another blocker: I'm able to achieve 100% matches on my own machine but GitHub Actions Playwright is always at 99%. We are now dealing with very subtle pixel discoloration problems and I don't have any good ideas.

Check out (just extract and open the `html` file) the latest test report here: 
[test-report.zip](https://github.com/user-attachments/files/18103517/test-report.zip)

<hr />

Less Favorable Workarounds:

1. Set threshold to 99% and accept 1% CI buffer.
2. Keep looking for a solution, blocking other enhancements.
3. Feel free to suggest other better options.

Hear from you soon! K.H

KemingHe commented:
Oh and two more things:

1. The `Inter variable font` yields the best match with current production UI.
2. Fixed weight vs variable font doesn't affect the 99% screenshot match rate (tested).

KemingHe commented:
Hi @wei , `build.yml` is confirmed to pass when the threshold for UI is set to 99% and the threshold for preview output is set to 100%: https://github.com/KemingHe/contrib-socialify/actions/runs/12287044326

wei commented:
I agree 99% is perfectly reasonable for main ui. Reviewing the rest of the PR. 

p.s. for the future it's best to keep PRs single purposed. Adding contributing guide is a great addition and probably warrants its own PR 😃.  I know you added it because test and e2e information so we can keep in this PR for the time being~

Let me know if there is anything else outstanding from your end. 🫶

KemingHe commented:
Hi Wei, I saw that the latest Action failed for your repo `wei/socialify` while passing for my `KemingHe/contrib-socialify` with the only diff variable I can think of being the `${{ secrets.PAT }}` in `build.yaml`.

Could you confirm that `PAT` as `repo` scope?

### Ref

#### Failed Run (`wei/socialify`)

https://github.com/wei/socialify/actions/runs/12287045202/job/34290278010?pr=404

#### Success Run (`KemingHe/contrib-socialify`)

https://github.com/KemingHe/contrib-socialify/actions/runs/12287044326/job/34288338442

#### File Ref

https://github.com/KemingHe/contrib-socialify/blob/da3d35d36a79fcbb0ad744192b429ceebee69154/.github/workflows/build.yml#L42-L49

wei commented:
Ok, thanks for the details. I will confirm and rerun.

wei commented:
@KemingHe can you invite me as a collaborator? 

![image](https://github.com/user-attachments/assets/c88e7aa7-d18d-4755-9cb7-549913374dde)

I guess the checkbox from https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork doesn't work for some reason.

KemingHe commented:
Hi @wei, I just turn the `Allow edits and access to secrets by maintainers` checkbox off and on again, PLUS invited you as a collaborator to `KemingHe/contrib-sociaify`.

Could you try the standard way first and then accept the invitation? Might just be a small GitHub glitch.

wei commented:
No luck. Only worked after accepting invite.

wei commented:
Found the problem, need to set [`pull_request_target`](https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#pull_request_target) as trigger instead of just `pull_request` in order to use the secrets.

wei reviewed (approved):
Got all the tests to pass Hooray! 🥳 

Thanks for all the hard work @KemingHe 🙏
